ipaserver/install/ipa_cert_fix.py

```diff
@@ -0,0 +1,272 @@
+    description = "Renew expired certificates."
+
+    def validate_options(self):
+        super(IPACertFix, self).validate_options(needs_root=True)
+
+    def run(self):
+        if not is_ipa_configured():
+            print("IPA is not configured.")
+            return 2
+
+        if not cainstance.is_ca_installed_locally():
+            print("CA is not installed on this server.")
+            return 1
+
+        try:
+            ipautil.run(['pki-server', 'cert-fix', '--help'], raiseonerr=True)
+        except ipautil.CalledProcessError:
+            print(
+                "The 'pki-server cert-fix' command is not available; "
+                "cannot proceed."
+            )
+            return 1
+
+        api.bootstrap(in_server=True, confdir=paths.ETC_IPA)
+        api.finalize()
+        api.Backend.ldap2.connect()  # ensure DS is up
+
+        subject_base = dsinstance.DsInstance().find_subject_base()
+        if not subject_base:
+            raise RuntimeError("Cannot determine certificate subject base.")
+
+        ca_subject_dn = ca.lookup_ca_subject(api, subject_base)
+
+        now = datetime.datetime.now() + datetime.timedelta(weeks=2)
+        certs, extra_certs = expired_certs(now)
+
+        if not certs and not extra_certs:
+            print("Nothing to do.")
+            return 0
+
+        print(msg)
+
+        print_intentions(certs, extra_certs)
+
+        response = ipautil.user_input('Enter "yes" to proceed')
+        if response.lower() != 'yes':
+            print("Not proceeding.")
+            return 0
+        print("Proceeding.")
+
+        run_cert_fix(certs, extra_certs)
+
+        replicate_dogtag_certs(subject_base, ca_subject_dn, certs)
+        install_ipa_certs(subject_base, ca_subject_dn, extra_certs)
+
+        if any(x != 'sslserver' for x in certs) \
+                or any(x[0] is IPACertType.IPARA for x in extra_certs):
+            # we renewed a "shared" certificate, therefore we must
+            # become the renewal master
+            print("Becoming renewal master.")
+            cainstance.CAInstance().set_renewal_master()
+
+        ipautil.run(['ipactl', 'restart'], raiseonerr=True)
+
+
+def expired_certs(now):
+    return expired_dogtag_certs(now), expired_ipa_certs(now)
+
+
+def expired_dogtag_certs(now):
+    """
+    Determine which Dogtag certs are expired, or close to expiry.
+
+    Return a list of (cert_id, cert) pairs.
+
+    """
+    certs = []
+    db = NSSDatabase(nssdir=paths.PKI_TOMCAT_ALIAS_DIR)
+
+    for certid, nickname in [
+        ('sslserver', 'Server-Cert cert-pki-ca'),
+        ('subsystem', 'subsystemCert cert-pki-ca'),
+        ('ca_ocsp_signing', 'ocspSigningCert cert-pki-ca'),
+        ('ca_audit_signing', 'auditSigningCert cert-pki-ca'),
+        ('kra_transport', 'transportCert cert-pki-kra'),
+        ('kra_storage', 'storageCert cert-pki-kra'),
+        ('kra_audit_signing', 'auditSigningCert cert-pki-kra'),
+    ]:
+        cert = db.get_cert(nickname)
```

Comment:
KRA certs should be optional. Otherwise, if KRA is not installed, ipa-cert-fix will exit on error with "Failed to get transportCert cert-pki-kra".